test: Drop `x` modifier in `fsbridge::fopen` call for MinGW builds

hebasto opened the pull request:
The MinGW-w64 toolchain links executables to the old msvcrt C Runtime Library that does not support the `x` modifier for the [`_wfopen()`](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/fopen-wfopen?view=msvc-170) function.

Fixes https://github.com/bitcoin/bitcoin/issues/29014.

hebasto commented:
> > Nevertheless, the second commit introduces an assertion in the AutoFile constructor.
> 
> Can you explain why this change makes sense? `AutoFile` already checks for nullptr.

The added assertion makes the issue with the unsupported `x` modifier explicit:

> The `streams_tests/xor_file` test will fail without the first commit.

fanquake commented:
Concept NACK. I don't think we should remove test coverage from all other platforms, just to accomodate Windows. In the worse case this should just have a Windows paths with the modified modifiers, with documentation explaining why?

hebasto commented:
> Concept NACK. I don't think we should remove test coverage from all other platforms, just to accomodate Windows.

It is not about Windows as a platform. It is about Windows users who download the Bitcoin Core release binaries. We should not treat them in a discriminatory way.

hebasto commented:
> It is already explicit and will fail on master without the first commit:
> 
> ```
> unknown location(0): fatal error: in "streams_tests/xor_file": std::ios_base::failure[abi:cxx11]: AutoFile::write: file handle is nullptr: iostream error
> ```

On the master branch test passes on my system (Ubuntu 22.04):
```
$ ./src/test/test_bitcoin.exe -t streams_tests/xor_file
Running 1 test case...
00f0:err:msvcrt:msvcrt_get_flags incorrect mode flag: x

*** No errors detected
```

fanquake commented:
> We should not treat them in a discriminatory way.

I don't see how fixing the tests in the way I've suggested would be discriminatory. It would make them work on Windows while retaining the current test for all other platforms.

hebasto commented:
> Can you add exact steps to reproduce?
```
$ cat /etc/os-release | head -1
PRETTY_NAME="Ubuntu 22.04.3 LTS"
$ x86_64-w64-mingw32-g++-posix --version
x86_64-w64-mingw32-g++-posix (GCC) 10-posix 20220113
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

$ make -C depends -j $(nproc) HOST=x86_64-w64-mingw32 NO_QT=1
$ ./autogen.sh && ./configure CONFIG_SITE=$PWD/depends/x86_64-w64-mingw32/share/config.site
$ make -j $(nproc) -C src test/test_bitcoin.exe
$ ./src/test/test_bitcoin.exe -t streams_tests/xor_file
Running 1 test case...

*** No errors detected
```

hebasto committed:
test: Drop `x` modifier in `fsbridge::fopen` call for mingw builds

The MinGW-w64 toolchain links executables to the old msvcrt C Runtime
Library that does not support the `x` modifier for the _wfopen()
function.

hebasto commented:
> > We should not treat them in a discriminatory way.
> 
> I don't see how fixing the tests in the way I've suggested would be discriminatory. It would make them work on Windows while retaining the current test for all other platforms.

Thank you! Your suggestion has been implemented.

hebasto commented:
> > > I also tried this locally, and the file is not a nullptr.
> > 
> > 
> > I agree. That is how it works on Linux in the Wine environment.
> > On Windows, being linked to `msvcrt.dll`, it fails.
> 
> Yes, either the file is a nullptr, in which case the test already fails, or it is not, in which case the test passes. In any case, the added assert in the dropped commit isn't needed here or elsewhere, and doesn't change the outcome here.

You are right. That commit was my mistake.

FWIW, the [`_wfopen_s`](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/fopen-s-wfopen-s?view=msvc-170) function reports an error for the `x` modifier even in Wine.

fanquake commented on src/test/streams_tests.cpp:
```diff
@@ -29,7 +29,18 @@ BOOST_AUTO_TEST_CASE(xor_file)
+        // function. It is possible to link to the new UCRT library instead.
+        // However, the most easiest way to do it is to use a new GCC with
+        // the -mcrtdll=ucrt option, which has been available since the commit
+        // https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=453cb585f0f8673a5d69d1b420ffd4b3f53aca00
```
I think this comment could just be something like `Our usage of mingw-w64 and the msvcrt runtime does not support the x modifier for the _wfopen()`. The other info seems a bit overkiil, I'm also not sure if it's correct. From a quick look, if you choose the runtime you want, when you compile mingw-w64, there's no need for any GCC flag.

fanquake replied:
> If the new runtime is available in guix, I'd say it would be better to just use it, instead of patching the code. Or is there some downside I am missing?

I agree that we should move to the new runtime. I assume the downside is possibly dropping support for Windows 7, but I don't see why that would be a problem.

hebasto replied:
> if you choose the runtime you want, when you compile mingw-w64

Yes, there is `--with-default-msvcrt=ucrt` configuration option.

fanquake replied:
Yea. So I don't see why anything GCC related is required, or needs to be mentioned here.

fanquake commented:
> FWIW, the [_wfopen_s](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/fopen-s-wfopen-s?view=msvc-170) function reports an error for the x modifier even in Wine.

So it reports an error but the unit tests don't fail? Is this another (different bug)?

hebasto commented:
> > FWIW, the [_wfopen_s](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/fopen-s-wfopen-s?view=msvc-170) function reports an error for the x modifier even in Wine.
> 
> So it reports an error but the unit tests don't fail? Is this another (different bug)?

We use the `_wfopen`. Microsoft docs [suggests](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/fopen-wfopen?view=msvc-170):
> More-secure versions of these functions that perform more parameter validation and return error codes are available; see [fopen_s, _wfopen_s](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/fopen-s-wfopen-s?view=msvc-170).

So, the `_wfopen_s` is an alternative, which is not used now.

fanquake commented:
I think the documentation should still be adjusted (https://github.com/bitcoin/bitcoin/pull/29357#discussion_r1475960179). No point merging what I still think are incorrect docs, especially if the change is going to be backported. In any case, our unit test code also seems like the wrong place to put overly verbose documentaiton about GCC configuration options and the various windows runtimes.

hebasto commented:
> I think the documentation should still be adjusted ([#29357 (comment)](https://github.com/bitcoin/bitcoin/pull/29357#discussion_r1475960179)).

Sorry, it skipped my attention. Fixed now.

fanquake reviewed (approved):
ACK d2fe90571e98e02617682561ea82acb1e2647827 - the plan here should still be to migrate to the newer windows runtime.

pablomartin4btc reviewed (commented):
post-merge ACK d2fe90571e98e02617682561ea82acb1e2647827

I've [tested](https://github.com/bitcoin/bitcoin/issues/29014#issuecomment-1858387980) this fix earlier on the related issue but missed the PR.

Safay99 reviewed (commented):
1. بیبی

Safay99 commented:
[
/](url)
